File: core/core.net/source/core.net/IO/BinaryWriter/BinaryWriter.cs
//
// Copyright (c) 2015. SeongJun Park.
//
// Distributed under the MIT License.
//

using System;
using System.Text;
using System.Runtime.InteropServices;
using System.Collections;
using System.Reflection;
using System.IO;

namespace Useless.IO
{
	public class BinaryWriter : IDisposable
	{
		class StreamProxy
		{
			private IStream m_uselessStream = null;
			private Stream m_stream = null;

			public StreamProxy( IStream stream )
			{
				m_uselessStream = stream;
			}

			public StreamProxy( Stream stream )
			{
				m_stream = stream;
			}

			public void Dispose()
			{
				if( m_uselessStream != null )
				{
					m_uselessStream.Dispose();
				}
				else if( m_stream != null )
				{
					m_stream.Dispose();
				}
			}

			public object GetStream()
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream;
				}
				else
				{
					return m_stream;
				}
			}

			public int Seek( int offset, SeekDir way )
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream.Seek( offset, way );
				}
				else
				{
					return ( int )m_stream.Seek( offset, ( SeekOrigin )way );
				}
			}

			public int GetPosition()
			{
				if( m_uselessStream != null )
				{

					return m_uselessStream.Position;
				}
				else
				{
					return ( int )m_stream.Position;
				}
			}

			public int Write( byte[] buffer, int count )
			{
				if( m_uselessStream != null )
				{
					return m_uselessStream.Write( buffer, count );
				}
				else
				{
					m_stream.Write( buffer, 0, count );
					return count;
				}
			}
		}

		private StreamProxy m_stream = null;
		private bool m_mustBeDispose = false;

		public BinaryWriter()
		{
		}

		public BinaryWriter( IStream stream )
		{
			if( stream == null || !stream.CanWrite )
			{
				throw new ArgumentException( "stream" );
			}

			m_stream = new StreamProxy( stream );
		}

		public BinaryWriter( Stream stream )
		{
			if( stream == null || !stream.CanWrite )
			{
				throw new ArgumentException( "stream" );
			}

			m_stream = new StreamProxy( stream );
		}

		public BinaryWriter( int size, bool growable = true )
		{
			if( growable )
			{
				m_stream = new StreamProxy( new DynamicStream( size ) );
			}
			else
			{
				m_stream = new StreamProxy( new FixedStream( size ) );
			}

			m_mustBeDispose = true;
		}

		public BinaryWriter( byte[] address, int size )
		{
			m_stream = new StreamProxy( new MemoryStream( address, size ) );
			m_mustBeDispose = true;
		}
		
		public BinaryWriter( String path, System.IO.FileMode openmode )
		{
			m_stream = new StreamProxy( new FileStream( path, openmode ) );
			m_mustBeDispose = true;
		}

		public void Dispose()
		{
			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = null;
		}

		public void SetStream( IStream stream )
		{
			if( stream == null || !stream.CanWrite )
			{
				throw new ArgumentException( "stream" );
			}

			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = new StreamProxy( stream );
			m_mustBeDispose = false;
		}

		public void SetStream( Stream stream )
		{
			if( stream == null || !stream.CanWrite )
			{
				throw new ArgumentException( "stream" );
			}

			if( m_stream != null && m_mustBeDispose )
			{
				m_stream.Dispose();
			}

			m_stream = new StreamProxy( stream );
			m_mustBeDispose = false;
		}

		public object GetStream()
		{
			return m_stream.GetStream();
		}

		public int Seek( int offset, SeekDir way )
		{
			return m_stream.Seek( offset, way );
		}

		public int GetPosition()
		{
			return m_stream.GetPosition();
		}

		public void Write( byte[] buffer, int count )
		{
			if( m_stream == null )
			{
				throw new ObjectDisposedException( "stream" );
			}

			m_stream.Write( buffer, count );
		}

		public void Write( Type type, object val )
		{
			if( TypeTraits.IsArithmetic.Invoke( type ) || TypeTraits.IsChar.Invoke( type ) )
			{
				Write( Utility.BitConverter.GetBytes( type, val ), Marshal.SizeOf( type ) );
			}
			else if( TypeTraits.IsString.Invoke( type ) )
			{
				String stringVal = ( String )val;
				int count = Encoding.UTF8.GetByteCount( stringVal );
                WriteU32( ( uint )count );
				Write( Encoding.UTF8.GetBytes( stringVal ), count );
			}
			else if( type.IsArray )
			{
				Type element_type = type.GetElementType();

				Array arrayVal = ( Array )val;
				WriteU32( ( uint )arrayVal.Length );
				foreach( object data in arrayVal )
				{
					Write( element_type, data );
				}
			}
			else if( val is IDictionary )
			{
				Type key_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];
				Type value_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 1 ];

				IDictionary dictionaryVal = ( IDictionary )val;
				WriteU32( ( uint )dictionaryVal.Count );
				IDictionaryEnumerator enumerator = dictionaryVal.GetEnumerator();
				while( enumerator.MoveNext() )
				{
					Write( key_type, enumerator.Key );
					Write( value_type, enumerator.Value );
				}
            }
			else if( val is ICollection )
			{
				Type element_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];

				ICollection collectionVal = ( ICollection )val;
				WriteU32( ( uint )collectionVal.Count );
				foreach( object data in collectionVal )
				{
					Write( element_type, data );
				}
			}
			else if( val is IEnumerable )
			{
				Type element_type = val.GetType().GetTypeInfo().GenericTypeArguments[ 0 ];

				IEnumerable enumerableVal = ( IEnumerable )val;
				int count = 0;
				foreach( object data in enumerableVal )
				{
					++count;
				}
				WriteU32( ( uint )count );

				foreach( object data in enumerableVal )
				{
					Write( element_type, data );
				}
			}
			else
			{
				ISerializable serializable = val as ISerializable;
				if( serializable == null )
				{
					throw new ObjectSerializeException();
				}

				BinaryWriter refThis = this;
				serializable.Serialize( ref refThis );
			}
		}

		public void Write<T>( T val )
		{
			object temp = val;
			Write( typeof( T ), temp );
		}

		public void WriteBoolean( bool val )
		{
			Write( val );
		}

		public void WriteS8( sbyte val )
		{
			Write( val );
		}

		public void WriteU8( byte val )
		{
			Write( val );
		}

		public void WriteS16( short val )
		{
			Write( val );
		}

		public void WriteU16( ushort val )
		{
			Write( val );
		}

		public void WriteS32( int val )
		{
			Write( val );
		}

		public void WriteU32( uint val )
		{
			Write( val );
		}

		public void WriteS64( long val )
		{
			Write( val );
		}

		public void WriteU64( ulong val )
		{
			Write( val );
		}

		public void WriteF32( float val )
		{
			Write( val );
		}

		public void WriteF64( double val )
		{
			Write( val );
		}

		public void WriteString( String val )
		{
			Write( val );
		}
	}
}

File: core/core.java/source/core.java/io/src/useless/io/BinaryWriter.java
/*
 * Copyright (c) 2015. SeongJun Park.
 *
 * Distributed under the MIT License.
 */

package useless.io;

import useless.io.exception.*;
import useless.type_traits.*;
import useless.utility.*;
import java.io.UnsupportedEncodingException;
import java.util.*;

public class BinaryWriter
{
	private StreamBase m_stream = null;

	public BinaryWriter()
	{

	}

	public BinaryWriter( StreamBase stream )
	{
		open( stream );
	}

	public BinaryWriter( int size, boolean growable )
	{
		open( size, growable );
	}

	public BinaryWriter( int size )
	{
		open( size );
	}

	public BinaryWriter( byte[] address, int size )
	{
		open( address, size );
	}

	public void open( StreamBase stream )
	{
		if( stream == null || !stream.canWrite() )
		{
			throw new IllegalArgumentException( "stream" );
		}

		m_stream = stream;
	}

	public void open( int size, boolean growable )
	{
		if( growable )
		{
			m_stream = new DynamicStream( size );
		}
		else
		{
			m_stream = new FixedStream( size );
		}
	}

	public void open( int size )
	{
		open( size, true );
	}

	public void open( byte[] address, int size )
	{
		m_stream = new MemoryStream( address, size );
	}

	public StreamBase getStream()
	{
		return m_stream;
	}

	public void write( byte[] buffer, int count )
	{
		if( m_stream == null )
		{
			throw new NullPointerException( "stream" );
		}

		int write = m_stream.write( buffer, count );
		if( write != count )
		{
			throw new WriteFailedException();
		}
	}

	public void write( Object val ) throws UnsupportedEncodingException, ObjectSerializeException
	{
		Class<?> type = val.getClass();
		if( IsArithmetic.invoke( type ) )
		{
			if( type == Boolean.class || type == boolean.class )
			{
				writeBoolean( ( Boolean ) val );
			}
			else if( type == Byte.class || type == byte.class )
			{
				writeS8( ( Byte ) val );
			}
			else if( type == Short.class || type == short.class )
			{
				writeS16( ( Short ) val );
			}
			else if( type == Integer.class || type == int.class )
			{
				writeS32( ( Integer ) val );
			}
			else if( type == Long.class || type == long.class )
			{
				writeS64( ( Long ) val );
			}
			else if( type == Float.class || type == float.class )
			{
				writeF32( ( Float ) val );
			}
			else if( type == Double.class || type == double.class )
			{
				writeF64( ( Double ) val );
			}
			else
			{
				throw new UnsupportedOperationException();
			}
		}
		else if( IsChar.invoke( type ) )
		{
			write( BitConverter.getBytes( ( Character )val ), 2 );
		}
		else if( IsString.invoke( type ) )
		{
			writeString( ( String ) val );
		}
		else if( type.isArray() )
		{
			writeArray( type, val );
		}
		else
		{
			if( val instanceof Collection<?> )
			{
				Collection<?> collectionVal = ( Collection<?> )val;
				int count = collectionVal.size();
				writeU32( ( long ) count );
				for( Object temp : collectionVal )
				{
					write( temp );
				}
			}
			else if( val instanceof Map<?, ?> )
			{
				Map<?, ?> mapVal = ( Map<?, ?> )val;
				int count = mapVal.size();
				writeU32( ( long ) count );
				Iterator<?> iterator = mapVal.entrySet().iterator();
				while( iterator.hasNext() )
				{
					Map.Entry temp = ( Map.Entry )iterator.next();
					write( temp.getKey() );
					write( temp.getValue() );
				}
			}
			else if( val instanceof Serializable )
			{
				( ( Serializable )val ).serialize( this );
			}
			else
			{
				throw new ObjectSerializeException();
			}
		}
	}

	public void writeBoolean( boolean val )
	{
		write( BitConverter.getBytes( val ), 1 );
	}

	public void writeS8( byte val )
	{
		write( BitConverter.getBytes( val ), 1 );
	}

	public void writeU8( short val )
	{
		write( BitConverter.getBytes_U8( val ), 1 );
	}

	public void writeS16( short val )
	{
		write( BitConverter.getBytes( val ), 2 );
	}

	public void writeU16( int val )
	{
		write( BitConverter.getBytes_U16( val ), 2 );
	}

	public void writeS32( int val )
	{
		write( BitConverter.getBytes( val ), 4 );
	}

	public void writeU32( long val )
	{
		write( BitConverter.getBytes_U32( val ), 4 );
	}

	public void writeS64( long val )
	{
		write( BitConverter.getBytes( val ), 8 );
	}

	public void writeF32( float val )
	{
		write( BitConverter.getBytes( val ), 4 );
	}

	public void writeF64( double val )
	{
		write( BitConverter.getBytes( val ), 8 );
	}

	public void writeString( String val ) throws UnsupportedEncodingException
	{
		String stringVal = ( String )val;
		byte[] stringBytes = stringVal.getBytes( "UTF-8" );
		int count = stringBytes.length;
		writeU32( ( long ) count );
		write( stringBytes, count );
	}

	public void writeArray( Class<?> type, Object val ) throws UnsupportedEncodingException, ObjectSerializeException
	{
		Class<?> element_type = type.getComponentType();
		if( IsArithmetic.invoke( element_type ) )
		{
			if( element_type == Boolean.class || element_type == boolean.class )
			{
				boolean[] ba = ( boolean[] )val;
				writeU32( ( long )ba.length );
				for( boolean temp : ba )
				{
					writeBoolean( temp );
				}
			}
			else if( element_type == Byte.class || element_type == byte.class )
			{
				byte[] ba = ( byte[] )val;
				writeU32( ( long )ba.length );
				for( byte temp : ba )
				{
					writeS8( temp );
				}
			}
			else if( element_type == Short.class || element_type == short.class )
			{
				short[] sa = ( short[] )val;
				writeU32( ( long )sa.length );
				for( short temp : sa )
				{
					writeS16( temp );
				}
			}
			else if( element_type == Integer.class || element_type == int.class )
			{
				int[] ia = ( int[] )val;
				writeU32( ( long )ia.length );
				for( int temp : ia )
				{
					writeS32( temp );
				}
			}
			else if( element_type == Long.class || element_type == long.class )
			{
				long[] la = ( long[] )val;
				writeU32( ( long )la.length );
				for( long temp : la )
				{
					writeS64( temp );
				}
			}
			else if( element_type == Float.class || element_type == float.class )
			{
				float[] fa = ( float[] )val;
				writeU32( ( long )fa.length );
				for( float temp : fa )
				{
					writeF32( temp );
				}
			}
			else if( element_type == Double.class || element_type == double.class )
			{
				double[] da = ( double[] )val;
				writeU32( ( long )da.length );
				for( double temp : da )
				{
					writeF64( temp );
				}
			}
			else
			{
				throw new UnsupportedOperationException();
			}
		}
		else if( IsChar.invoke( element_type ) )
		{
			char[] ca = ( char[] )val;
			writeU32( ( long )ca.length );
			for( char temp : ca )
			{
				write( BitConverter.getBytes( temp ), 2 );
			}
		}
		else
		{
			Object[] arrayVal = ( Object[] )val;
			writeU32( ( long )arrayVal.length );

			for( Object obj : arrayVal )
			{
				write( obj );
			}
		}
	}
}
